### ISSUE TITLE
Suggestion: Also include parameter default values in $PSBoundParameters

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#3285]
Currently, a script/function's parameter is only considered _bound_ if an actual value was supplied by the _caller_.

Notably, this means that parameter _default values_ are _not_ reflected in  `$PSBoundParameters`, which decreases its usefulness when the intent is to pass _all_ parameter values through to another command via splatting (`@PSBoundParameters`).

For a real-world example and workarounds, see [this Stack Overflow](http://stackoverflow.com/q/42677229/45375) question.

**Comment by Jaykul** [comment]
Yeah, I've written a function `Get-ParameterValue` several times for this purpose. Particularly when splatting to another function it's awkward to have to deal with it by hand. It would be nice if there was a `$PSAllParameters` or something.

**Comment by BrucePay** [comment]
There is a strong reason not to change the current behavior: the presence of a member in this collection tells you that the parameter was actually specified in the command call i.e. that the formal parameter was bound hence the name "PSBoundParameters". You'll have to add a new variable if you want different semantics.

**Comment by lzybkr** [comment]
Adding a new property is reasonable though - we have `BoundPositionally` list of strings - we could a similar `BoundDefaultValue` list.

**Comment by True-Demon** [comment]
I realize this issue has long since been closed, but for those of us who want a realistic solution, I have implemented some rather simple boilerplate code for it.

```
foreach($key in $MyInvocation.MyCommand.Parameters.Keys){
    $value = Get-Variable $key -ValueOnly -EA SilentlyContinue
    if($value -and !$PSBoundParameters.ContainsKey($key)) {$PSBoundParameters[$key] = $value}
}
```
Or if you want a more concise method or would rather simply cast them into a new variable to use independent of $PSBoundParameters...

`$PSAllParameters = $MyInvocation.MyCommand.Parameters`
In both cases, you can use these for splatting in the same way.

```
MyNormalFunction @PSBoundParameters
MyEverythingFunction @PSAllParameters
```

**Comment by mklement0** [comment]
As for your workarounds:

The first one is great, but I don't think the second one will work, though, because `$MyInvocation.MyCommand.Parameters` contains only metadata, not values, from what I can tell.

**Comment by Jaykul** [comment]
@mklement0 That's why he calls get-variable 😉 

Here's mine, which I mentioned earlier. It's basically the same thing, except that it merges with PSBoundParameters (and is in function form, so you can dot-source it, and includes tests to make _sure_ you dot-sourced it)

https://gist.github.com/Jaykul/c623c45d32329cadac9e7622937e1b76

**Comment by mklement0** [comment]
@Jaykul:

> That's why he calls get-variable

Yes, in the first snippet, which indeed works (as stated - though I'd use `$null -ne $value` rather than just `$value` as a slightly more robust conditional to infer the presence of a value from; it too modifies `$PSBoundParameters` directly).

However, my comment was about the second snippet, which I interpret to make the claim that `$MyInvocation.MyCommand.Parameters` would essentially already give us what this issues requests as an enhancement, which is not the case; specifically, I was referring to:

> you can use the native $MyInvocation.MyCommand.Parameters, which contains all parameter keys & values, bound, null, default, or otherwise

My point was that `$MyInvocation.MyCommand.Parameters` seemingly contains just metadata, not values (neither bound nor default ones).

Thanks for sharing your Gist, which is a nicely packaged solution.
Sounds like we have viable workarounds now.
Let's hope that we soon won't need them anymore.

**Comment by True-Demon** [comment]
You're right, I made the mistake of assuming before I posted that. I've omitted that since it doesn't apply in order to prevent leading others astray.

**Comment by stinos** [comment]
@True-Demon @Jaykul did you never get into trouble with these workarounds in combination with dynamic parameters? Problem is `Get-Variable` gets any variable so in case of dynamic parameters it won't find their default because those parameters aren't available with their name in the scope, so instead it looks further so if there happens to be another, unrelated, variable with the same name in the scope it will use that. E.g.

```
$foo = 1

function Bar {
dynamicparam(
  # Create "foo" dynamic here with default 2
)
# Oops, this returns 1, no matter what was passed.
Get-Variable "foo"
}
```

**Comment by mklement0** [comment]
@stinos, you can use `Get-Variable -Scope Local` to limit looking for variables in the current scope:

```powershell
foreach ($key in $MyInvocation.MyCommand.Parameters.Keys) {
  $value = if ($PSBoundParameters.ContainsKey($key)) { $PSBoundParameters[$key] } 
           else { Get-Variable -Scope Local -ErrorAction Ignore -ValueOnly $key }
}
```

**Comment by stinos** [comment]
@mklement0 thanks for posting that, I've actually been using something like that myself but apparently forgot to report back here.

**Comment by agowa** [comment]
Just fixed a bug in a script that used:
```
if ('Debug' -in $PSBoundParameters.Keys) {}
```
This needs to be fixed or at least we need warnings from psscriptanalyzer (and vscode)...
This at first looks innocent, but considering the `-Debug:$false` semantic also exists it clearly shows the issue with this. It's probably not just limited to this, I quickly googled for how best practices about how to use it and found code like this:
```
if ($PSBoundParameters.Debug) {}
```
which works as long as `-Debug:$true` or `-Debug:$false`, but causes an exception when absent, it's non terminating, but still. If changing this parameter to defaulting to all values being populated and set to false is not feasible right now, can we at least get this 2nd case to also work in StrictMode and return $false instead of this exception?:
```

Exception             :
    Type        : System.Management.Automation.PropertyNotFoundException
    ErrorRecord :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : The property 'Debug' cannot be found on this object. Verify that the property exists.
            HResult : -2146233087
        CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : PropertyNotFoundStrict
        InvocationInfo        :
            ScriptLineNumber : 1
            OffsetInLine     : 43
            HistoryId        : -1
            Line             : function t {[cmdletbinding()]param(); if ($PSBoundParameters.Debug) {'true'} else
{'false'}}
            PositionMessage  : At line:1 char:43
                               + … ion t {[cmdletbinding()]param(); if ($PSBoundParameters.Debug) {'true …
                               +                                        ~~~~~~~~~~~~~~~~~~~~~~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at t, <No file>: line 1
                                at <ScriptBlock>, <No file>: line 1
    TargetSite  : System.Object CallSite.Target(System.Runtime.CompilerServices.Closure,
System.Runtime.CompilerServices.CallSite, System.Management.Automation.PSBoundParametersDictionary)
    Message     : The property 'Debug' cannot be found on this object. Verify that the property exists.
    Data        : System.Collections.ListDictionaryInternal
    Source      : Anonymously Hosted DynamicMethods Assembly
    HResult     : -2146233087
    StackTrace  :
   at CallSite.Target(Closure, CallSite, PSBoundParametersDictionary)
   at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : NotSpecified: (:) [], PropertyNotFoundException
FullyQualifiedErrorId : PropertyNotFoundStrict
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 43
    HistoryId        : -1
    Line             : function t {[cmdletbinding()]param(); if ($PSBoundParameters.Debug) {'true'} else {'false'}}
    PositionMessage  : At line:1 char:43
                       + … ion t {[cmdletbinding()]param(); if ($PSBoundParameters.Debug) {'true …
                       +                                        ~~~~~~~~~~~~~~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at t, <No file>: line 1
                        at <ScriptBlock>, <No file>: line 1

```

Testcode:
```
Set-StrictMode -Version latest
function t {[cmdletbinding()]param(); if ($PSBoundParameters.Debug) {'true'} else {'false'}; 'continued'}
t
# throws exception but continues execution AFTER the if block, at least as long as $ErrorActionPreference is set to the default of 'Continue'
# And outputs just the string 'continued'
Set-StrictMode -Off
t
# Outputs both, the string 'false' and 'continued'
```

**Comment by Jaykul** [comment]
@agowa wrote:
> Just fixed a bug in a script that used:
> 
> ```pwsh
> if ('Debug' -in $PSBoundParameters.Keys) {}
> ```

For Debug, you should normally rely on `$DebugPreference`. You shouldn't be testing for the `-Debug` switch explicitly, because it's a state (i.e. if -Debug is set on the caller, it doesn't need to be repeated to each nested command). 

Having said that, for switch parameters in the general case, you may need to test like this:
```pwsh
if ($PSBoundParameters.ContainsKey('Debug') -and $PSBoundParameters['Debug']) {}
```

**Comment by agowa** [comment]
@Jaykul That wasn't my code I found that "gem" in the code of coworkers...

**Comment by Jaykul** [comment]
I'm not sure what you're suggesting. Those switches are legacy behavior it would be dangerous to mess with.

Since the beginning, advanced functions in PowerShell have had `-Verbose` and `-Debug` switches (along with `-ErrorAction` and `-WarningAction` and the newer `-InformationAction` and -ProgressAction`) to control the preference variables ... you can't just _change_ what they put in PSBoundParameters, because you'd break everything that already exists.

I mean, we _recently_ finally got `-Debug` to set `$DebugPreference` to "Continue" instead of "Inquire" (which was mostly useless), but changing it to `-DebugAction` and `-VerboseAction` would break everything, and having the common parameters set _extra variables_ would _also_ break everything (since you'd be reserving those variables as system variables after having them be usable for all these years).

**Comment by agowa** [comment]
Another thought. If we say Switches from CommonParameters are actually intended to be the way they currently are with three states `$null`, `$true`, `$false`, I.E. a nullable switches, can we change their type definition to reflect this `[System.Nullable[System.Management.Automation.SwitchParameter]` more easily without breakage?

Because currently the Parameters property of both `Get-Command foo` and the related `Get-Item Function:/foo` show it as a regular `[System.Management.Automation.SwitchParameter]`, which can only have two states `$true` and `$false`.

**Comment by mubed** [comment]
> Adding a new property is reasonable though - we have `BoundPositionally` list of strings - we could a similar `BoundDefaultValue` list.

I think this is the most effortless and elegant way. I would include the arguments set via default value and list them in something like `BoundByDefaultValue` for those who really need to know, where the value came from, which would really be a rare case in my opinion.

The cleanest and most OOP way would be to touch the dictionary objects and add a property to them like `IsSetByDefaultValue` (see https://github.com/PowerShell/PowerShell/issues/21263), but that would be a more difficult implementation.

**Comment by mklement0** [comment]
While the idea of adding a `.BoundByDefaultValue` may still be worthwhile (though I agree that needing to know that is probably uncommon), it would still the make the use case that motivated creation of this issue cumbersome, which is  to have a _variable_ that can be used _as-is_ - possibly following modification of its value - to _splat_ all parameter values, _whether bound by the caller or by default value_, say with `@PSParametersWithValue`.

While creating a parallel dictionary to `$PSBoundParameters` may be undesirable from a performance and resource-duplication perspective, perhaps a new `$PSParametersWithValue` variable can be implemented _like a property that constructs its value dynamically_, on first access. 
(I assume this is feasible with behind-the-scenes magic for _automatic_ variables, but I don't know.)

That way, the extra effort would only be expended in code that actively references this variable.

**Comment by KL-Sven** [comment]
+1
